src/uct/ib/dc/dc_mlx5_ep.c

```diff
@@ -1404,18 +1380,43 @@ ucs_status_t uct_dc_mlx5_ep_check_fc(uct_dc_mlx5_iface_t *iface,
         return UCS_OK;
     }
 
-    status = uct_rc_fc_ctrl(&ep->super.super, UCT_RC_EP_FLAG_FC_HARD_REQ,
-                            NULL);
+    now = ucs_get_time();
+    it  = kh_put(uct_dc_mlx5_fc_hash, &iface->tx.fc_hash, (uint64_t)ep,
+                 &ret);
+    if (ucs_unlikely(ret == UCS_KH_PUT_FAILED)) {
+        ucs_error("failed to create hash entry for fc hard req");
+        status = UCS_ERR_NO_MEMORY;
+        goto out;
+    } else if (ret == UCS_KH_PUT_KEY_PRESENT) {
+        fc_entry = &kh_value(&iface->tx.fc_hash, it);
+
+        /* Do not resend FC request if timeout is not reached, or window is
+         * still not empty */
+        if (ucs_likely((ep->fc.fc_wnd > 0) ||
+                       ((now - fc_entry->send_time) <
+                                iface->tx.fc_hard_req_timeout))) {
+            goto out_set_status;
+        }
+    } else {
+        ucs_assert(ep->fc.fc_wnd ==
+                           iface->super.super.config.fc_hard_thresh);
+        fc_entry = &kh_value(&iface->tx.fc_hash, it);
+    }
```

Comment:
why is it moved from `uct_dc_mlx5_ep_fc_hard_req_send`? Imo we could try to resend grant for all type of packets, not only for `ep_check`

Reply:
`uct_dc_mlx5_ep_check_fc()` is used for all type of packets and for EP_CHECK operation.
it is moved here to have two functions:
- uct_dc_mlx5_ep_fc_hard_req_send()
- uct_dc_mlx5_ep_fc_pure_grant_send()
instead of `uct_rc_fc_ctrl()`

Reply:
the current version of this, where hash access is outside of uct_dc_mlx5_ep_fc_hard_req_send, seems better to me

Reply:
> the current version of this, where hash access is outside of uct_dc_mlx5_ep_fc_hard_req_send, seems better to me

ok, good